# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ayushmanlohani/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [12]:
!git clone https://github.com/ayushmanlohani/flyrank-ml-internship.git
%cd flyrank-ml-internship

Cloning into 'flyrank-ml-internship'...
remote: Enumerating objects: 122, done.
remote: Counting objects: 100% (122/122), done.
remote: Compressing objects: 100% (78/78), done.
remote: Total 122 (delta 35), reused 97 (delta 28), pack-reused 0 (from 0)
Receiving objects: 100% (122/122), 1.86 MiB | 10.00 MiB/s, done.
Resolving deltas: 100% (35/35), done.
/content/flyrank-ml-internship/flyrank-ml-internship/flyrank-ml-internship


## 1. My lane is content freshness.

*I want to investigate whether older or less recently updated pages are more likely to be declining in search performance.*

In [13]:
import pandas as pd

# Load dataset
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# Create the boolean label for declining pages
df['is_declining_label'] = df['trend_direction'] == 'down'

# Calculate and display the mean metrics
metrics = df.groupby('is_declining_label')[['content_age_days', 'days_since_last_update']].mean()
print("Mean values for content age and freshness:")
print(metrics)

Mean values for content age and freshness:
                    content_age_days  days_since_last_update
is_declining_label                                          
False                     279.829451               42.372543
True                      236.178637               49.245788


## 2. The question: decision, action, cost of a wrong call

- **What decision does this improve?** It improves the decision of **which older or declining pages should be refreshed/updated first** by a content editor.
- **Who acts on it?** The **Content Editorial Team / Editors** who will prioritize and update the flagged pages.
- **What does a wrong recommendation cost?**
  - *False Positive:* Wasted editor hours spent rewriting pages that didn't need updates.
  - *False Negative:* Missing a declining page, leading to lost traffic, impressions, and revenue.

In [14]:
print("Declining pages:", df["is_declining_label"].sum())
print("Total pages:", len(df))
print("Declining rate:", df["is_declining_label"].mean())

Declining pages: 16262
Total pages: 30000
Declining rate: 0.5420666666666667


## 3. Quick look at the data (2-3 real numbers)

From our initial run on the anonymized data, we found:
1. **Freshness Impact:** Pages that are declining in performance (`trend_direction == 'down'`) have not been updated for **49.2 days** on average, compared to **42.4 days** for stable or growing pages.
2. **Overall Age:** Declining pages have a mean overall content age of **236.2 days**, whereas non-declining pages have a mean age of **279.8 days**.

In [15]:
print(df.groupby("is_declining_label")[["content_age_days", "days_since_last_update"]].mean())

                    content_age_days  days_since_last_update
is_declining_label                                          
False                     279.829451               42.372543
True                      236.178637               49.245788


## 4. Careful words: what I can and can't claim

- **What I can claim:** We can claim **directional and decision-support relationships** (e.g., *"Pages that haven't been updated for over 49 days are directionally more likely to show performance decline than those updated within 42 days"*).
- **What I can't claim:** We cannot claim **causal proof** (e.g., we cannot prove that the lack of updates is the sole cause of decline) or claim that we are **'predicting Google'** or its search ranking algorithms directly.

In [16]:
print("Mean days since update:")
print(df.groupby("is_declining_label")["days_since_last_update"].mean())

print("\nMean content age:")
print(df.groupby("is_declining_label")["content_age_days"].mean())

Mean days since update:
is_declining_label
False    42.372543
True     49.245788
Name: days_since_last_update, dtype: float64

Mean content age:
is_declining_label
False    279.829451
True     236.178637
Name: content_age_days, dtype: float64


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.